### PR TITLE
[docs] Refactor s3 upload in multiple stages for reliabilty

### PR DESCRIPTION
**Pull request by byCedric**
# Why

This fixes two main things.

1. **Reliability of the deployments**
Unfortunately, it looks like `$ aws s3 sync` isn't doing exactly what we would expect. [It uses some methods to determine if a file needs to be updated](https://github.com/aws/aws-cli/issues/3273#issuecomment-643436849), which causes some issues on our end. Resulting in either JS or HTML files not being uploaded or updated. This change mostly uses `$ aws s3 cp` to do what `$ aws s3 sync` currently does. It's still used to clean up the old deployment files, but not for uploading the new one.

2. **Staged rollout**
Instead of copying everything alphabetically, this change also introduces stages. These are explained in the deployment script itself, so I don't think I need to highlight it. TL;DR; it tries to upload stuff, before breaking the previous deployment.

One thing to note is that I use `::group::` and `::endgroup::` to create collapsible items in Github Actions (making it more debuggable). It also adds `--no-progress` to keep the logs readable as well.

# How

See files.

# Test Plan

You can see [a test/staging deployment over here](https://cedrics-expo.s3.eu-central-1.amazonaws.com/index.html), using this exact script. Over here, you can see the [full output of the logs](https://cedrics-expo.s3.eu-central-1.amazonaws.com/deployment-log.txt).

### CODE REVIEW

**Comment by byCedric** [comment]
@ide I think the staged rollout was exactly the thing you mentioned, right?

File: docs/deploy.sh [hunk 1]
```diff
@@ -11,14 +11,58 @@ if [ ! -d "$target" ]; then
...
+  "$target/static/" \
+  "s3://${bucket}/static/"
+echo "::endgroup::"
+
+echo "::group::[3/5] Overwrite all HTML and other files, not located in \`_next/**\` or \`static/**\` folder"
+aws s3 cp \
+  --no-progress \
+  --recursive \
+  --metadata-directive REPLACE \
+  --cache-control "public,max-age=31536000,immutable" \
```
Comment (ide) [inline docs/deploy.sh:52]: I think this is going to cause very serious issues, don't think we want caching, this should be no-cache

File: docs/deploy.sh [hunk 2]
```diff
@@ -11,14 +11,58 @@ if [ ! -d "$target" ]; then
...
+  "$target/_next/" \
+  "s3://${bucket}/_next/"
+echo "::endgroup::"
+
+echo "::group::[2/5] Upload new asset files in \`static/**\` folder"
 aws s3 cp \
+  --no-progress \
   --recursive \
   --metadata-directive REPLACE \
   --cache-control "public,max-age=31536000,immutable" \
```
Comment (ide) [inline docs/deploy.sh:42]: The caching for `/static/` will likely cause serious issues, this should be a 1-hour cache maximum (that is, let the site be buggy for 1 hour max... if we want zero bugs then we should turn off caching). Only cache content-addressed files permanently.

**Review by ide — CHANGES_REQUESTED** [review]
Does `aws s3 cp` avoid copying files if their checksums have not changed? We don't change our large images and videos on most builds. The reason we use `sync` is to simulate that behavior of rsync and other similar tools.

**Comment by byCedric** [comment]
> Does `aws s3 cp` avoid copying files if their checksums have not changed? We don't change our large images and videos on most builds. The reason we use `sync` is to simulate that behavior of rsync and other similar tools.

No it doesn't check for the checksum, it uploads the file directly and overwrites it in S3. The main problem is are the HTML files, sometimes they were skipped even though they refer to different JS checksums.

We could also opt-in to use `$ aws s3 sync` for `static/**` files. That could save some bandwidth because the assets don't change that often, and I don't think we encountered the HTML-mismatch with these files. For `_next/**` I don't think it has a lot of effect, every export has a different checksum somehow and that results in the files being different anyways.

**Comment by ide** [comment]
OK. Maybe let's try `aws s3 sync` for `_next/**` (pr `cp` would work as you mentioned) and `static/**` and use `cp` for the HTML files, followed with a final `sync` to delete unused files. The important thing is to upload dependencies before dependents. And ideally we would delete old dependents (HTML) before old dependencies (JS).

File: docs/deploy.sh [hunk 3]
```diff
@@ -11,14 +11,42 @@ if [ ! -d "$target" ]; then
...
+# see: https://github.com/aws/aws-cli/issues/3273#issuecomment-643436849
+
+echo "::group::[1/4] Upload JS files in \`_next/**\` folder"
 aws s3 cp \
+  --no-progress \
   --recursive \
-  --metadata-directive REPLACE \
-  --cache-control "public,max-age=31536000,immutable" \
-  "s3://${bucket}/_next/static/" \
-  "s3://${bucket}/_next/static/"
```
Comment (byCedric) [inline docs/deploy.sh:21]: This static folder contains the JS files of the pages. Looking at your comments James, I'm not sure if we should keep this. It doesn't contain assets like images or videos.

**Comment by byCedric** [comment]
> And ideally we would delete old dependents (HTML) before old dependencies (JS).

When we `$ aws s3 cp` the HTML files, we actually "force-replace" these old dependents (HTML). Running `$ aws s3 sync --delete` will delete the outdated dependencies (JS). 

It would probably look like this:
1. Run `$ aws s3 sync` on `_next/**` and `static/**` only (without delete) 
2. Run `$ aws s3 cp` to force-overwriting HTML files.
3. Run `$ aws s3 sync --delete` on the full folder, for both assets and cleaning up the JS files.

**Comment by byCedric** [comment]
All right, refactored it based on input and tested it again on my own bucket:
> You can see [a test/staging deployment over here](https://cedrics-expo.s3.eu-central-1.amazonaws.com/index.html), using this exact script. Over here, you can see the [full output of the logs](https://cedrics-expo.s3.eu-central-1.amazonaws.com/deployment-log.txt).

File: docs/deploy.sh [hunk 4]
```diff
@@ -11,14 +11,44 @@ if [ ! -d "$target" ]; then
...
+# To keep the previous website up and running, we deploy it using these steps.
+#   1.  Sync JS/assets dependencies in \`_next/**\` and \`static/**\` folder
+#      > Uploads the new generated JS and asset files (stored in hashed folders to avoid collision with older deployments)
+#   2. Overwrite HTML dependents, not located in \`_next/**\` or \`static/**\` folder
+#      > Force overwrite of all HTML files to make sure we use the latest one
+#   3. Sync assets and clean up outdated files from previous deployments
+#   4. Add custom redirects
 
+# Due to a bug with `aws s3 sync` we need to copy everything first instead of syncing
+# see: https://github.com/aws/aws-cli/issues/3273#issuecomment-643436849
```
Comment (ide) [inline docs/deploy.sh:23]: Move this comment down to the cp section

**Review by ide — APPROVED** [review]

**Pull request merged**